### PR TITLE
Support non-keyword dimensions as routing fields in TSDB

### PR DESCRIPTION
**Pull request by kkrik-es**
This is a follow-up on https://github.com/elastic/elasticsearch/pull/103648, lifting the limitation that the pass-through object can contain only keyword fields if it's marked as time_series_dimension. In particular, fields included in the routing path no more need to be keywords.

#106080 included all required changes around routing. This change builds on top of it, including changes in mappings. In particular, it streamlines mapping eligibility as dimension fields.

Related to https://github.com/elastic/elasticsearch/issues/103567

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java [hunk 1]
```diff
@@ -237,12 +241,16 @@ public void collectSearchShards(String routing, IntConsumer consumer) {
     public static class ExtractFromSource extends IndexRouting {
         private final Predicate<String> isRoutingPath;
         private final XContentParserConfiguration parserConfig;
+        private final boolean indexTracksRoutingId;
+
+        private int routingId = 0;
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java:246]: I think in some cases we reuse this instance and if then  `indexShard(...)` gets invoked by multiple callers then this gets tricky around what routing id will be read in postProcess.

File: server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java [hunk 2]
```diff
@@ -253,7 +261,11 @@ public boolean matchesField(String fieldName) {
         }
 
         @Override
-        public void process(IndexRequest indexRequest) {}
+        public void postProcess(IndexRequest indexRequest) {
+            if (indexRequest.id() == null && indexTracksRoutingId) {
+                indexRequest.id(TimeSeriesRoutingIdFieldMapper.encode(routingId));
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java:266]: I think it is cleaner to add a routingHash field to `IndexRequest` instead of reusing the id property here.

File: server/src/main/java/org/elasticsearch/action/DocWriteRequest.java [hunk 3]
```diff
@@ -161,7 +161,12 @@ public interface DocWriteRequest<T> extends IndicesRequest, Accountable {
     /**
      * Finalize the request before executing or routing it.
      */
-    void process(IndexRouting indexRouting);
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/action/DocWriteRequest.java:164]: Let's undo this change to `DocWriteRequest` and instead add IndexRequest as a parameter to `IndexRouting#indexShard(...)` method. I think this should be a simpler change.

 Or maybe add a `Consumer` parameter that then would provide the routing hash to IndexRequest.

File: server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesRoutingIdFieldMapper.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+import org.elasticsearch.index.fielddata.plain.SortedOrdinalsIndexFieldData;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.script.field.DelegateDocValuesField;
+import org.elasticsearch.search.aggregations.support.CoreValuesSourceType;
+
+import java.util.Base64;
+import java.util.Collections;
+
+/**
+ * Mapper for the {@code _seq_no} field.
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesRoutingIdFieldMapper.java:29]: Update the java docs? :)

File: server/src/main/java/org/elasticsearch/index/mapper/IdLoader.java [hunk 5]
```diff
@@ -66,21 +66,24 @@ final class TsIdLoader implements IdLoader {
...
-                // Routing field must always be keyword fields, so it is ok to use SortedSetDocValues directly here.
-                SortedSetDocValues dv = DocValues.getSortedSet(reader, routingField);
-                for (int i = 0; i < docIdsInLeaf.length; i++) {
-                    int docId = docIdsInLeaf[i];
-                    var builder = builders[i];
-                    if (dv.advanceExact(docId)) {
-                        for (int j = 0; j < dv.docValueCount(); j++) {
-                            BytesRef routingValue = dv.lookupOrd(dv.nextOrd());
-                            builder.addMatching(routingField, routingValue);
+                for (String routingField : routingPaths) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/IdLoader.java:76]: Can we just read the routing bit from the `routing_id` field? Or is this bwc logic?

File: server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesRoutingIdFieldMapper.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+ * This mapper also manages the primary term field, which has no ES named
+ * equivalent. The primary term is only used during collision after receiving
+ * identical seq# values for two document copies. The primary term is stored as
+ * a doc value field without being indexed, since it is only intended for use
+ * as a key-value lookup.
+
+ */
+public class TimeSeriesRoutingIdFieldMapper extends MetadataFieldMapper {
+
+    public static final String NAME = "_routing_id";
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesRoutingIdFieldMapper.java:44]: I think we should refer to routing id to routing hash instead? It isn't really an id and it is only used as a hash.
In that case the `TimeSeriesRoutingIdFieldMapper` should renamed to `TimeSeriesRoutingHashFieldMapper`, `_routing_id` should become `_routing_hash` and similar change in other places.

File: server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesRoutingIdFieldMapper.java (hunk 4 above)
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesRoutingIdFieldMapper.java:29]: Done.

File: server/src/main/java/org/elasticsearch/index/mapper/IdLoader.java (hunk 5 above)
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/IdLoader.java:76]: Yeah this is the old logic, for bwc.

File: server/src/main/java/org/elasticsearch/action/DocWriteRequest.java (hunk 3 above)
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/action/DocWriteRequest.java]: Yeah a consumer looks better, good point.

File: server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java (hunk 2 above)
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java]: Went with using the routing param, seems to have no side effects.

**Comment by kkrik-es** [comment]
This is getting pretty long, I'm forking the `_ts_routing_hash` and routing logic to #106080. Once that's in, this one will be used just to update the mappings.

File: server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java (hunk 1 above)
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java]: Right, switched to use a Consumer.

**Comment by kkrik-es** [comment]
The forked PR is in, back to this. Changes are now limited to mappings, mostly.

File: server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java [hunk 7]
```diff
@@ -195,6 +195,20 @@ public boolean isDimension() {
         return false;
     }
 
+    /**
+     * @return true if field can be marked as a dimension field
+     */
+    public boolean supportsDimension() {
```
[the remark this answers is not in the document]
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:201]: There are two cases we need to check:
 - Can this field type be a dimension? This changes error messages and behavior under PassThroughObjectMapper.
 - Is a particular mapper a dimension? This has very specific semantics, dimension vs label vs metric.

File: server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java [hunk 8]
```diff
@@ -195,6 +195,20 @@ public boolean isDimension() {
...
+        return false;
+    }
+
+    /**
+     * @return true if field has script values.
+     */
+    public boolean hasScriptValues() {
```
[the remark this answers is not in the document]
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:208]: This was lifted from `KeywordFieldMapper`. I think it's used only when a dimension is used in the routing path - it was already the case for keywords and it'll be applied to other types that this PR enables for such use.

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java [hunk 9]
```diff
@@ -1415,6 +1415,19 @@ private static boolean isDeprecatedParameter(String propName, IndexVersion index
...
 
+    public abstract static class DimensionBuilder extends Builder {
+
+        protected boolean isDimension = false;
+
+        public DimensionBuilder(String name) {
+            super(name);
+        }
+
+        void setDimension() {
```
[the remark this answers is not in the document]
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:1426]: Hm that would be misleading, we want to mark that the mapper to be built will actually be a dimension, not that the builder is enabled to build a dimension (it is, de factor).

I really don't like the name though. This class creates objects that *can* be dimensions but are not necessarily. Any suggestions?

File: server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:201]: Looking the usages of `supportsDimension(...)` in production code it is only used to return a slightly different error message in the `validateMatchedRoutingPath(...)` method. That feels a bit of overkill to me? I think the `if (isDimension() == false) {` in `validateMatchedRoutingPath(...)` method is sufficient? So I think we don't need the `supportsDimension()` method?

File: server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:208]: I see, this makes sense.

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:1426]: It would be nice if we can move `Parameter<Boolean> dimension` into this builder, but that looks like a bigger refactoring to me. 

So passthrough field mapper invokes `setDimension()` here and then other field mappers (keyword, number, etc.) can become dimensions.  Maybe this method should be called `inheritDimension()`?

[the remark this answers is not in the document]
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java]: Alright, switched to `inheritDimensionParameterFromParentObject`, more verbose for sure.

**Review by martijnvg — APPROVED** [review]
LGTM
